base/array.jl

```diff
@@ -1063,24 +1076,188 @@ function setindex!(A::Array{T}, X::Array{T}, c::Colon) where T
     return A
 end
 
-# efficiently grow an array
+# Pick new memory size for efficiently growing an array
+# TODO: This should know about the size of our GC pools
+# Specifically we are wasting ~10% of memory for small arrays
+# by not picking memory sizes that max out a GC pool
+function overallocation(maxsize)
+    maxsize < 8 && return 8;
+    # compute maxsize = maxsize + 4*maxsize^(7/8) + maxsize/8
+    # for small n, we grow faster than O(n)
+    # for large n, we grow at O(n/8)
+    # and as we reach O(memory) for memory>>1MB,
+    # this means we end by adding about 10% of memory each time
+    exp2 = sizeof(maxsize) * 8 - Core.Intrinsics.ctlz_int(maxsize)
+    maxsize += (1 << div(exp2 * 7, 8)) * 4 + div(maxsize, 8)
+    return maxsize
+end
+
+function array_new_memory(mem::Memory{T}, newlen::Int) where T
+    if ccall(:jl_mem_owner, Any, (Memory{T},), mem) isa String
+        # if data is in a String, keep it that way
+        # TODO: use jl_gc_expand_string(oldstr, newlen)?
+        str = _string_n(newlen)
+        return ccall(:jl_string_to_genericmemory, Memory{T}, (Any,), str)
+    else
+        # TODO: when implimented, this should use a memory growing call
+        mem = Memory{T}(undef, newlen)
+        return mem
+    end
+end
 
-_growbeg!(a::Vector, delta::Integer) =
-    ccall(:jl_array_grow_beg, Cvoid, (Any, UInt), a, delta)
-_growend!(a::Vector, delta::Integer) =
-    ccall(:jl_array_grow_end, Cvoid, (Any, UInt), a, delta)
-_growat!(a::Vector, i::Integer, delta::Integer) =
-    ccall(:jl_array_grow_at, Cvoid, (Any, Int, UInt), a, i - 1, delta)
+function _growbeg!(a::Vector, delta::Integer)
+    delta = Int(delta)
+    delta == 0 && return # avoid attempting to index off the end
+    delta >= 0 || throw(ArgumentError("grow requires delta >= 0"))
+    ref = a.ref
+    mem = ref.mem
+    len = length(a)
+    offset = memoffset(ref)
+    newlen = len + delta
+    a.size = (newlen,)
+    # if offset is far enough advanced to fit data in existing memory without copying
+    if delta <= offset
+        a.ref = MemoryRef(ref, 1 - delta, false)
+    else
+        @noinline (function()
```

Comment:
Can someone describe the use of this `@noinline (function() ...)()` pattern here? What benefit it is providing to the compiler?

Reply:
this is for outlining the code. Specifically we want there to be a function call here to make the fast-path faster.